File: LeetCodeProblemsGo/LeetCode907SumOfSubarrayMinimums/LeetCode907SumOfSubarrayMinimums.go
package LeetCode907SumOfSubarrayMinimums

func sumSubarrayMins(arr []int) int {
	var modulo int64 = 1e9 + 7
	length := len(arr)
	leftFirstSmaller := make([]int, length)
	rightFirstSamaller := make([]int, length)
	monotonique := &Stack{}
	monotonique.initialise()
	monotonique.push([]int{arr[0], 0})
	leftFirstSmaller[0] = -1
	for i := 1; i < length; i++ {
		for (!monotonique.isEmpty()) && monotonique.peek()[0] > arr[i] {
			monotonique.pop()
		}
		if monotonique.isEmpty() {
			leftFirstSmaller[i] = -1
			monotonique.push([]int{arr[i], i})
			continue
		}
		leftFirstSmaller[i] = monotonique.peek()[1]
		monotonique.push([]int{arr[i], i})
	}
	monotonique.initialise()
	rightFirstSamaller[length-1] = length
	monotonique.push([]int{arr[length-1], length - 1})
	for i := length - 2; i >= 0; i-- {
		for (!monotonique.isEmpty()) && monotonique.peek()[0] >= arr[i] {
			monotonique.pop()
		}
		if monotonique.isEmpty() {
			rightFirstSamaller[i] = length
			monotonique.push([]int{arr[i], i})
			continue
		}
		rightFirstSamaller[i] = monotonique.peek()[1]
		monotonique.push([]int{arr[i], i})
	}
	var result int64 = 0
	var delta int64 = 0
	for i := 0; i < length; i++ {
		delta = int64(arr[i])
		delta *= int64(i - leftFirstSmaller[i])
		delta %= modulo
		delta *= int64(rightFirstSamaller[i] - i)
		delta %= modulo
		result += delta
		result %= modulo
	}
	return int(result)
}

type Stack struct {
	items [][]int
}

func (s *Stack) initialise() {
	s.items = [][]int{}
}
func (s *Stack) isEmpty() bool {
	return len(s.items) == 0

}
func (s *Stack) peek() []int {
	return s.items[len(s.items)-1]
}
func (s *Stack) pop() []int {
	top := s.peek()
	s.items = s.items[:len(s.items)-1]
	return top
}
func (s *Stack) push(pair []int) {
	s.items = append(s.items, pair)
}

File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode907SumOfSubarrayMinimums.cpp
#include "LeetCode907SumOfSubarrayMinimums.h"

#include <iostream>
#include <queue>
#include <stack>
using namespace std;

int LeetCode907SumOfSubarrayMinimums::sumSubarrayMins(vector<int>& arr)
{
    int modulo=1e9+7;
    stack<pair<int, int>> monotonicStack{};
    int length=arr.size();
    vector<int> leftFirstLesserIndex(length,-1);
    vector<int> rightFirstLesserIndex(length,length);
    monotonicStack.emplace(arr[0],0);
    pair<int,int> top{};
    for(int i=1;i<length;i++)
    {
        while ((!monotonicStack.empty()) && monotonicStack.top().first>=arr[i])
        {
            monotonicStack.pop();
        }
        if (monotonicStack.empty())
        {
            monotonicStack.emplace(arr[i],i);
            leftFirstLesserIndex[i]=-1;
            continue;
        }
        leftFirstLesserIndex[i]=monotonicStack.top().second;
        monotonicStack.emplace(arr[i],i);
    }
    monotonicStack={};
    monotonicStack.emplace(arr[length-1],length-1);
    for (int i=length-2;i>=0;i--)
    {
        while ((!monotonicStack.empty())&&monotonicStack.top().first> arr[i])
        {
            monotonicStack.pop();
        }
        if (monotonicStack.empty())
        {
            monotonicStack.emplace(arr[i],i);
            rightFirstLesserIndex[i]=length;
            continue;
        }
        rightFirstLesserIndex[i]=monotonicStack.top().second;
        monotonicStack.emplace(arr[i],i);
    }
    long long result{};
    long long delta{};
    for (int i=0;i<length;i++)
    {
        delta=arr[i];
        delta*= i-(leftFirstLesserIndex[i]+1)+1;
        delta%=modulo;
        delta*=rightFirstLesserIndex[i]-i;
        delta%=modulo;
        result+=delta;
        result%=modulo;
    }
    return result;
}

void LeetCode907SumOfSubarrayMinimums::Run()
{
    vector<int> arr={71,55,82,55};
        // {3,1,2,4};
    cout<<sumSubarrayMins(arr)<<endl;
}
